File: src/queue.ts
import { Road } from "./io.js";

export type Vehicle = {
    id: string;
    endRoad: Road;
    next: Vehicle | null;
};

export class Queue {
    private head: Vehicle | null;
    private tail: Vehicle | null;
    private count: number;

    constructor() {
        this.head = null;
        this.tail = null;
        this.count = 0;
    }

    enqueue(id: string, endRoad: Road): void {
        const v: Vehicle = {
            id: id,
            endRoad: endRoad,
            next: null,
        };

        if (this.tail) {
            this.tail.next = v;
        } else {
            this.head = v;
        }
        this.tail = v;

        this.count++;
    }

    dequeue(): Vehicle | null {
        if (!this.head) {
            return null;
        }

        const v = this.head;
        this.head = v.next;

        if (!this.head) {
            this.tail = null;
        }
        this.count--;

        return v;
    }

    getCount(): number {
        return this.count;
    }
}

File: src/queue.c
#include "queue.h"

#ifndef EMBEDDED
#include <stdlib.h>
#include <string.h>

void enqueue(Queue* q, const char* name) {
    Vehicle* v = malloc(sizeof(Vehicle));
    v->name = strdup(name);
    v->next = NULL;

    if (q->tail) {
        q->tail->next = v;
    } else {
        q->head = v;
    }
    q->tail = v;

    q->count++;
}

// Free the returned pointer after being done with it
Vehicle* dequeue(Queue* q) {
    if (!q->head) {
        return NULL;
    }

    Vehicle* v = q->head;
    q->head = v->next;

    if (!q->head) {
        q->tail = NULL;
    }
    q->count--;

    return v;
}
#endif

#ifdef EMBEDDED
// Without heap (static array)
#endif
